### DISCUSSION TITLE
[Spring Graphql] Knowing which fields are requested by a query/subscription?

### DISCUSSION THREAD

**Discussion by nielsbasjes** [graphql-java/graphql-java#3444] [Q&A] [answered]
I'm trying to make a GraphQL interface on top of the data coming from a very slow physical device (I have both a solar converter and a heatpump that both only speak [modbus](https://en.wikipedia.org/wiki/Modbus)).

The root problem I have is that these devices have many fields available (my solarconverter has a few hundred fields) and there is a limit on modbus of a maximum of only 125 fields per call. On some of my devices such a call sometimes takes almost a second.

To allow getting the data as often as possible I want to only fetch the fields from the real device for which there is an actual need in either a query or subscription.

In a related project (not modbus, so I get the full record in all cases) I have used the Spring Graphql library and used a [QueryMapping](https://github.com/nielsbasjes/dsmr-tools/blob/ecf66a764cf81c3522e752f3d6c5a062422e77ac/dsmr-graphql/src/main/java/nl/basjes/dsmr/graphql/GraphQlRequestHandler.java#L49-L54) and a [SubscriptionMapping](https://github.com/nielsbasjes/dsmr-tools/blob/ecf66a764cf81c3522e752f3d6c5a062422e77ac/dsmr-graphql/src/main/java/nl/basjes/dsmr/graphql/GraphQlRequestHandler.java#L65-L74) to implement the query and subscription. 

This works fine but I do not see a way to get the data I think I need.

For the modbus case I see a need to have 
- A way to know which fields are requested in both the query and subscription scenarios.
- Have hooks/handlers for starting and stopping the subscription that can register/unregister the need for the queried fields towards the (slow) backend.

The only example code I have found so far that does something like this was this ~7 year old  [StockTickerWebSocket](https://github.com/graphql-java/graphql-java-subscription-example/blob/master/src/main/java/com/graphql/example/http/StockTickerWebSocket.java) example. 
This does however use something completely different than Spring-graphql.

What I'm looking for a some guidance going into the right direction.

- Can what I need be done with spring-graphql?
- If it cannot be done with spring-graphql (yet) then is the example I found still the best way forward?
- If this is not the best way; what is?

In all of the above: Any example/javadoc/code link that points me towards how to do this is highly appreciated.

Thanks.

**Comment by Andras-Csanyi** [comment, marked as answer]
The bottom of [this](https://www.graphql-java.com/blog/deep-dive-data-fetcher-results) article is about you need I believe. I tried it a few weeks ago in a Spring 3 app and the context includes the fields got queried.

**Reply by nielsbasjes** [reply]
Thanks. I have some reading to do :)

**Reply by nielsbasjes** [reply]
If anyone runs into this question and needs an example: Mine can be found here https://github.com/nielsbasjes/modbus-graphql/

From the schema defined by the actual data backend (in this case it is Modbus based) I generate the GraphQL schema ([here](https://github.com/nielsbasjes/modbus-graphql/blob/00adc5ea0a980cd8b1d56bef912bc736d77c9841/modbus-graphql-common/src/main/kotlin/nl/basjes/modbus/graphql/GraphQLSchemaModbusSchema.kt)) and at query/subscription time I make the mapping back using [this code](https://github.com/nielsbasjes/modbus-graphql/blob/00adc5ea0a980cd8b1d56bef912bc736d77c9841/modbus-graphql-common/src/main/kotlin/nl/basjes/modbus/graphql/SchemaDeviceGraphQLResolver.kt#L72-L82).
